Remove system-level outputs from storage and replace with demand component (follow-on to 631) (#666)
NREL/GreenHEART: docs/technology_models/storage_models_index.md

# Storage Models
Storage technologies input and output the 'Storage Commodity' (`commodity`) at different times. These technologies can be filled or charged, then unfilled or discharged at some later time. These models are usually constrained by two key model parameters: storage capacity and charge/discharge rate.

## Storage Performance Models
- [`StoragePerformanceModel`](#simple-generic-storage-performance)
- [`PySAMBatteryPerformanceModel`](#pysam-battery-performance)
- `StorageAutoSizingModel`

The following sections detail the inputs and outputs of the storage performance models.

```{note}
The inputs and outputs of storage performance models are generalized here for any commodity. If input and output names include the word `commodity`, the actual variable name would be the commodity defined for that storage model. For example, the `PySAMBatteryPerformanceModel` can only be used for the commodity  `electricity`. Therefore, the `commodity_in` input to the `PySAMBatteryPerformanceModel` is actually named `electricity_in`.
```

### Inputs
- `commodity_in`: commodity available to use for charging storage

> If using a **feedback control strategy** (this means that `tech_to_dispatch_connections` are defined in the plant configuration file for the storage technology), the control-related inputs for control to the storage performance:

Reviewer: ```suggestion
If using a **feedback control strategy** (this means that the controller received the actual storage state periodically), the control-related inputs for control to the storage performance include:
```

- `SOC`: the storage state-of-charge in units of either `percent` (values between 0 and 100) or `unitless` (values between 0 and 1)

The aggregated or summarized performance outputs are (single values):
- `storage_duration`: the storage capacity divided by the storage discharge rate. Units are time-based, such as `h` (hours).
- `rated_commodity_production`: the storage discharge rate in units of `commodity_rate_units`.
- `total_commodity_produced`: the summation of `commodity_out` over the simulation in units of `commodity_amount_units`. This value may be negative if the storage charges more than discharges.

The results that are output per-year of the `plant_life` are:
- `annual_commodity_produced`: each value is the `total_commodity_produced` scaled to a 1-year (8760 hours) simulation. This value may be negative if the storage charges more than discharges.
- `capacity_factor`: the storage capacity factor calculated as capacity factors are calculated in converter technologies, a ratio of the sum of *`total_commodity_produced`* to the discharge rate (or `rated_commodity_production`). This value may be negative if the storage charges more than discharges.

>     $$

Reviewer: just single "$". The extras are rendering around your equations.

>     CF = \frac{\sum_{t=0}^{n_{timesteps}}(\text{commodity_out})}{\text{discharge_rate}*n_{\text{timesteps}}}

Reviewer: Rather than time steps, it should be time.

    $$
- `standard_capacity_factor`: the storage capacity factor as defined by the [NLR ATB](https://atb.nrel.gov/electricity/2024b/utility-scale_battery_storage). The ratio of the total *commodity discharged* to the discharge rate (or `rated_commodity_production`). This value will always be greater than or equal to zero.
    $$

>     CF_{\text{standard}} = \frac{\sum_{t=0}^{n_{timesteps}}(\text{storage_commodity_discharge})}{\text{discharge_rate}*n_{\text{timesteps}}}

Reviewer: over time*rating, not timesteps*rating

The results that are output per-year of the `plant_life` are:
- `annual_commodity_produced`: each value is the `total_commodity_produced` scaled to a 1-year (8760 hours) simulation. This value may be negative if the storage charges more than discharges.
- `capacity_factor`: the storage capacity factor calculated as capacity factors are calculated in converter technologies, a ratio of the sum of *`total_commodity_produced`* to the discharge rate (or `rated_commodity_production`). This value may be negative if the storage charges more than discharges.

>     $
>     CF = \frac{\sum_{t=0}^{n_{time}}(\text{commodity_out})}{\text{discharge_rate}*n_{\text{time}}}
>     $

Reviewer: I don't think `n_time` is correct. I'm sorry I was not more specific.

```suggestion
    $
    CF=\frac{\sum_{t=0}^{n_{\text{time\_steps}}}(\text{commodity\_out}*dt)}{\text{discharge\_rating}*\text{total\_time}}
    $
```

Author: I think n_timesteps is correct. That's the variable we use in H2I so I think it makes sense as it was before.

Reviewer: I still disagree and would like to have the equations as I suggested. Please see my comment on the next equation.

Author: I added the dt in

Author: I think for consistency with the rest of the doc page, that using `n_timesteps` and `discharge_rate` are more appropriate variables in this equation. I added in `*dt` and the `\` to improve formatting for the underscores in text variables. I also don't know how the `total time` variable you're proposing is different than `n_timesteps`. Is it equal to `n_timesteps*dt`? (in which case, wouldn't the `dt` cancel out from the numerator and denominator ?)

Reviewer: Ultimately, we are going for total_energy_produced/total_nameplate_energy.  I think we should shoot for having energy in the numerator and denominator even if terms would technically cancel.

Reviewer: I'm fine with n_timesteps*dt

Author: heard - done!

Reviewer: I think this version would achieve what we need
```markdown
$
CF = \frac{\sum_{i=0}^{n_{\text{time\_steps}}}(\text{commodity\_out$_i$}*dt)}{\text{discharge\_rate}*n_{\text{time\_steps}}*dt}
$
```

Reviewer: I suggested switching _t to _i, but I can roll with where it is at now

Author: am I right that the only difference is the $\text{commodity}_i$ (I've been using $t$ though, so I'd make it $\text{commodity}_t$) instead of $\text{commodity}$ (it'd be commodity_out instead of commodity but GitHub doesn't like the underscores)?

If so - I just pushed up that change.

- `standard_capacity_factor`: the storage capacity factor as defined by the [NLR ATB](https://atb.nrel.gov/electricity/2024b/utility-scale_battery_storage). The ratio of the total *commodity discharged* to the discharge rate (or `rated_commodity_production`). This value will always be greater than or equal to zero.

>     $
>     CF_{\text{standard}} = \frac{\sum_{t=0}^{n_{time}}(\text{storage_commodity_discharge})}{\text{discharge_rate}*n_{\text{time}}}
>     $

Reviewer: ```suggestion
$
CF=\frac{\sum_{t=0}^{n_{\text{time\_steps}}}(\text{storage\_commodity\_discharge}*dt)}{\text{discharge\_rating}*\text{total\_time}}
$
```

Author: I don't think the dt stuff in the eqn necessarily important at this time. The storage performance models are not yet flexible to non-hourly dt - so I think we can keep it as-is. These are intended to just show the differences in the CF calculations.

Reviewer: I think we should write our math correctly, even if we don't have our code written properly. The dt in the code is implicit in code because the time steps are 1 hour, but dt is technically there.

Author: okay. I added in dt
